File: scripts/convert-images-to-webp.js
#!/usr/bin/env node
/**
 * PuniCodex — WebP conversion + picture markup updater.
 *
 * Scans raster images referenced by the public site, converts missing or
 * stale PNG/JPG siblings to WebP (quality 85), and updates templates + HTML
 * to use <picture> with WebP source + original fallback. Adds explicit
 * width/height attributes to reduce CLS and lazy-loads below-the-fold images.
 *
 * Run: node scripts/convert-images-to-webp.js
 */

const { spawnSync } = require('node:child_process');
const fs = require('node:fs');
const path = require('node:path');

const ROOT = path.resolve(__dirname, '..');

function writeFileWithRetry(filePath, data, encoding = 'utf8', retries = 5, delay = 100) {
  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      fs.writeFileSync(filePath, data, encoding);
      return;
    } catch (err) {
      const isTransient =
        err.code === 'EBUSY' ||
        err.code === 'EAGAIN' ||
        err.code === 'UNKNOWN' ||
        err.code === 'EPERM';
      if (attempt === retries || !isTransient) {
        throw err;
      }
      const ms = delay * attempt;
      console.warn(`  transient write error for ${filePath} (${err.code}), retrying in ${ms}ms...`);
      Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
    }
  }
}

function writeFileAtomic(filePath, data, encoding = 'utf8') {
  const tmpPath = `${filePath}.tmp.${process.pid}`;
  writeFileWithRetry(tmpPath, data, encoding);
  // The rename is just as lock-prone on Windows (AV scanners grab the fresh
  // .tmp file): retry with the same transient backoff as the write itself.
  for (let attempt = 1; attempt <= 5; attempt++) {
    try {
      fs.renameSync(tmpPath, filePath);
      return;
    } catch (err) {
      const isTransient =
        err.code === 'EBUSY' ||
        err.code === 'EAGAIN' ||
        err.code === 'UNKNOWN' ||
        err.code === 'EPERM';
      if (attempt === 5 || !isTransient) {
        try {
          fs.unlinkSync(tmpPath);
        } catch {}
        throw err;
      }
      const ms = 100 * attempt;
      console.warn(`  transient rename error for ${filePath} (${err.code}), retrying in ${ms}ms...`);
      Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
    }
  }
}

const LAZY_LOAD_CLASSES = new Set([
  'footer-logomark',
  'pantheon-mascot-img',
  'scholars-media-thumb',
  'gallery-lightbox-img',
  'creative-thumb',
]);

function runPython() {
  const pyPath = path.join(ROOT, 'scripts', 'convert-images-to-webp.py');
  const result = spawnSync(process.platform === 'win32' ? 'python' : 'python3', [pyPath], {
    cwd: ROOT,
    encoding: 'utf8',
    maxBuffer: 1024 * 1024,
  });

  if (result.error) {
    console.error('Python helper failed to spawn:', result.error.message);
    process.exit(1);
  }
  if (result.status !== 0) {
    console.error('Python helper exited with code', result.status);
    console.error(result.stderr || result.stdout);
    process.exit(result.status || 1);
  }

  try {
    return JSON.parse(result.stdout.split('\n').filter(Boolean).pop() || '{}');
  } catch {
    console.error('Could not parse Python helper output:', result.stdout);
    return { converted: [], count: 0 };
  }
}

function getImageDimensions(imgPath) {
  const base = path.basename(imgPath || '');
  // Site mascots / lockups / logomarks share identical canvas sizes.
  if (/mascot\.(png|jpe?g|webp)$/i.test(base)) return { width: 1024, height: 1536 };
  if (/logolockup\.(png|jpe?g|webp)$/i.test(base)) return { width: 1536, height: 1024 };
  if (/logomark\.(png|jpe?g|webp)$/i.test(base)) return { width: 1024, height: 1024 };
  // Public mascot directory.
  if (/assets\/images\/mascots\/[^/]+_mascot\.(png|webp)$/i.test(imgPath.replace(/\\/g, '/'))) {
    return { width: 1024, height: 1536 };
  }
  // OG default fallback.
  if (/og-default\.(png|webp)$/i.test(base)) return { width: 1200, height: 630 };

  // Fallback: read from disk for any other raster.
  if (imgPath && fs.existsSync(imgPath)) {
    const python = process.platform === 'win32' ? 'python' : 'python3';
    const script = `
import json, sys
from PIL import Image
with Image.open(sys.argv[1]) as im:
    print(json.dumps({"width": im.width, "height": im.height}))
`;
    const result = spawnSync(python, ['-c', script, imgPath], {
      encoding: 'utf8',
      maxBuffer: 64 * 1024,
    });
    try {
      return JSON.parse(result.stdout.trim().split('\n').pop() || '{}');
    } catch {
      return {};
    }
  }
  return {};
}

function resolveImagePath(htmlDir, src) {
  if (!src || src.startsWith('http') || src.startsWith('data:')) return null;
  const rel = src.startsWith('/') ? src.slice(1) : path.join(htmlDir, src);
  return path.resolve(ROOT, rel);
}

function normalizeAttributeOrder(tag, attrs) {
  // Preserve existing attributes while ensuring width/height/loading/decoding are present.
  let updated = tag;
  for (const [key, value] of Object.entries(attrs)) {
    const escaped = value.replace(/"/g, '&quot;');
    const regex = new RegExp(`\\b${key}="[^"]*"`, 'i');
    if (regex.test(updated)) {
      updated = updated.replace(regex, `${key}="${escaped}"`);
    } else {
      updated = updated.replace(/\s*\/?>$/, ` ${key}="${escaped}"$&`);
    }
  }
  return updated;
}

function updateImgTag(match, htmlDir, dimCache) {
  const fullTag = match[0];
  const classMatch = fullTag.match(/\bclass="([^"]+)"/);
  const classes = classMatch ? classMatch[1].split(/\s+/) : [];
  const srcMatch = fullTag.match(/\bsrc="([^"]+)"/);
  const src = srcMatch ? srcMatch[1] : null;

  const isLazyCandidate = classes.some((c) => LAZY_LOAD_CLASSES.has(c));

  const imgPath = resolveImagePath(htmlDir, src);
  const dims = imgPath ? getOrCacheDims(imgPath, dimCache) : {};

  const updates = {};
  if (dims.width && dims.height) {
    updates.width = String(dims.width);
    updates.height = String(dims.height);
  }
  if (isLazyCandidate && !/\bloading=/.test(fullTag)) {
    updates.loading = 'lazy';
  }
  if (isLazyCandidate && !/\bdecoding=/.test(fullTag)) {
    updates.decoding = 'async';
  }
  if (!updates.width) return fullTag;

  return normalizeAttributeOrder(fullTag, updates);
}

function getOrCacheDims(imgPath, cache) {
  if (!cache[imgPath]) {
    cache[imgPath] = getImageDimensions(imgPath);
  }
  return cache[imgPath];
}

function needsPictureWrap(src) {
  if (!src) return false;
  const ext = path.extname(src).toLowerCase();
  return ['.png', '.jpg', '.jpeg'].includes(ext);
}

function wrapImgInPicture(match) {
  const fullTag = match[0];
  const srcMatch = fullTag.match(/\bsrc="([^"]+)"/);
  const src = srcMatch ? srcMatch[1] : null;
  if (!needsPictureWrap(src)) return fullTag;

  const webpSrc = src.replace(/\.(png|jpe?g)$/i, '.webp');
  return `<picture><source srcset="${webpSrc}" type="image/webp">${fullTag}</picture>`;
}

function updatePictureBlock(block, htmlDir, dimCache) {
  // Update the inner <img> of an existing <picture> block.
  return block.replace(/<img\b[^>]*>/gi, (tag) => {
    return updateImgTag({ 0: tag, index: 0, input: block }, htmlDir, dimCache);
  });
}

function updateHtmlFile(filePath, dimCache) {
  const original = fs.readFileSync(filePath, 'utf8');
  const htmlDir =
    path
      .dirname(filePath)
      .replace(ROOT, '')
      .replace(/^[\\/]/, '') || '.';

  let updated = original;

  // 1. Update existing <picture>...</picture> blocks and replace with placeholders.
  const pictureRegex = /<picture\b[^>]*>.*?<\/picture>/gis;
  const pictureMap = new Map();
  let pictureIndex = 0;
  updated = updated.replace(pictureRegex, (block) => {
    const replacement = updatePictureBlock(block, htmlDir, dimCache);
    const key = `__PICTURE_PLACEHOLDER_${pictureIndex++}__`;
    pictureMap.set(key, replacement);
    return key;
  });

  // 2. Wrap standalone raster <img> tags in <picture> and update attributes.
  updated = updated.replace(/<img\b[^>]*>/gi, (tag) => {
    const wrapped = wrapImgInPicture({ 0: tag, index: 0, input: '' });
    const block = wrapped === tag ? tag : wrapped;
    return updatePictureBlock(block, htmlDir, dimCache);
  });

  // 3. Restore updated <picture> blocks.
  for (const [key, value] of pictureMap) {
    updated = updated.replace(key, value);
  }

  // 4. Replace og-default.svg with og-default.webp in meta / twitter / JSON-LD.
  updated = updated.replace(
    /(content=["'][^"']*)\/assets\/images\/og-default\.svg(["'])/gi,
    '$1/assets/images/og-default.webp$2'
  );
  updated = updated.replace(
    /"url":\s*"https:\/\/punicodex\.com\/assets\/images\/og-default\.svg"/gi,
    '"url": "https://punicodex.com/assets/images/og-default.webp"'
  );

  if (updated !== original) {
    writeFileAtomic(filePath, updated, 'utf8');
    return true;
  }
  return false;
}

function main() {
  console.log('▸ Running WebP conversion helper...');
  const result = runPython();
  if (result.error) {
    console.error('✗ Conversion failed:', result.error);
    process.exit(1);
  }
  console.log(`  Converted ${result.count || 0} image(s)`);

  const dimCache = {};
  const htmlDirs = [path.join(ROOT, 'templates', 'flagship'), path.join(ROOT, 'sites')];
  const rootHtmlFiles = [
    'index.html',
    'search.html',
    'oracle.html',
    'about/index.html',
    'appraise/index.html',
    'authenticity/index.html',
    'codex/index.html',
    'codex/anatomy-of-a-punycode-domain/index.html',
    'codex/building-the-temple/index.html',
    'codex/why-greek-accents-matter/index.html',
    'connections/index.html',
    'contact/index.html',
    'creatives/index.html',
    'game/index.html',
    'lexicon/index.html',
    'mobile/index.html',
    'pantheon/index.html',
    'privacy/index.html',
    'realms/index.html',
    'store/index.html',
    'tiers/index.html',
    'type/index.html',
  ];

  let updatedFiles = 0;

  console.log('▸ Updating templates + generated HTML...');

  for (const dir of htmlDirs) {
    if (!fs.existsSync(dir)) continue;
    const files = fs.readdirSync(dir, { recursive: true });
    let processed = 0;
    for (const rel of files) {
      const full = path.join(dir, rel);
      if (!full.endsWith('.html')) continue;
      if (!fs.statSync(full).isFile()) continue;
      if (updateHtmlFile(full, dimCache)) updatedFiles++;
      processed++;
      if (processed % 200 === 0) {
        Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 25);
      }
    }
  }

  for (const rel of rootHtmlFiles) {
    const full = path.join(ROOT, rel);
    if (fs.existsSync(full) && updateHtmlFile(full, dimCache)) updatedFiles++;
  }

  console.log(`  Updated ${updatedFiles} HTML file(s)`);
  console.log('✓ WebP conversion and markup update complete.');
}

main();

File: scripts/convert-images-to-webp.py
#!/usr/bin/env python3
"""
PuniCodex — WebP image conversion helper.

Converts raster images to WebP (quality 85) while preserving originals as
fallbacks. Also renders the OG default SVG to PNG/WebP.

This script is spawned by scripts/convert-images-to-webp.js. It prints JSON
metadata about converted files to stdout.
"""
import json
import os
import re
import sys
from pathlib import Path

try:
    from PIL import Image, ImageDraw, ImageFont
except ImportError as e:
    print(json.dumps({"error": f"Pillow not available: {e}"}), file=sys.stderr)
    sys.exit(1)

ROOT = Path(__file__).resolve().parent.parent
WEBP_QUALITY = 85


def draw_letter_spaced_text(draw, pos, text, font, fill, spacing):
    """Draw text with custom letter spacing; Pillow lacks native support."""
    x, y = pos
    total_width = sum(draw.textlength(ch, font=font) for ch in text)
    total_width += spacing * max(0, len(text) - 1)
    start_x = x - total_width / 2
    cursor = start_x
    for ch in text:
        draw.text((cursor, y), ch, fill=fill, font=font, anchor="lm")
        cursor += draw.textlength(ch, font=font) + spacing


def render_og_default():
    """Render assets/images/og-default.svg to PNG + WebP."""
    svg_path = ROOT / "assets" / "images" / "og-default.svg"
    png_path = ROOT / "assets" / "images" / "og-default.png"
    webp_path = ROOT / "assets" / "images" / "og-default.webp"

    if not svg_path.exists():
        return []

    if (
        png_path.exists()
        and webp_path.exists()
        and png_path.stat().st_mtime >= svg_path.stat().st_mtime
        and webp_path.stat().st_mtime >= svg_path.stat().st_mtime
    ):
        return []

    width, height = 1200, 630
    img = Image.new("RGB", (width, height), "#0A0A0A")
    draw = ImageDraw.Draw(img)

    # Gradient background (#0A0A0A -> #1A1200)
    pixels = img.load()
    for y in range(height):
        t = y / max(height - 1, 1)
        r = int(10 + (26 - 10) * t)
        g = int(10 + (18 - 10) * t)
        b = int(10 + 0 * t)
        for x in range(width):
            pixels[x, y] = (r, g, b)

    # Radial glow approximation (gold, 15% -> 0% opacity)
    import math

    glow = Image.new("RGBA", (width, height), (0, 0, 0, 0))
    cx, cy = width // 2, int(height * 0.4)
    max_r = int(min(width, height) * 0.5)
    glow_pixels = glow.load()
    for y in range(height):
        for x in range(width):
            d = math.hypot(x - cx, y - cy)
            if d < max_r:
                alpha = int(38 * (1 - d / max_r))  # 0.15 * 255 ≈ 38
                glow_pixels[x, y] = (212, 175, 55, alpha)
    img = Image.alpha_composite(img.convert("RGBA"), glow).convert("RGB")
    draw = ImageDraw.Draw(img)

    def load_font(size):
        candidates = [
            "C:/Windows/Fonts/georgia.ttf",
            "C:/Windows/Fonts/arial.ttf",
            "C:/Windows/Fonts/segoeui.ttf",
            "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
            "/System/Library/Fonts/Helvetica.ttc",
        ]
        for c in candidates:
            try:
                return ImageFont.truetype(c, size)
            except Exception:
                continue
        return ImageFont.load_default()

    font_title = load_font(120)
    font_subtitle = load_font(28)
    font_meta = load_font(16)

    draw_letter_spaced_text(
        draw, (600, 260), "PuniCodex", font_title, "#D4AF37", 8
    )
    draw_letter_spaced_text(
        draw, (600, 340), "THE UNICODE PANTHEON", font_subtitle, "#A0A0A0", 12
    )
    draw.line([(500, 390), (700, 390)], fill="#D4AF37", width=1)
    draw_letter_spaced_text(
        draw,
        (600, 430),
        "26 DOMAINS · 24 ARCHETYPES · ONE MISSION",
        font_meta,
        "#666666",
        2,
    )

    png_path.parent.mkdir(parents=True, exist_ok=True)
    img.save(png_path, "PNG")
    img.save(webp_path, "WEBP", quality=WEBP_QUALITY, method=6)
    return [str(png_path.relative_to(ROOT)), str(webp_path.relative_to(ROOT))]


def convert_image(src: Path):
    """Convert a single raster image to WebP if missing or stale."""
    ext = src.suffix.lower()
    if ext not in (".png", ".jpg", ".jpeg"):
        return None

    dst = src.with_suffix(".webp")
    if dst.exists() and dst.stat().st_mtime >= src.stat().st_mtime:
        return None

    try:
        with Image.open(src) as im:
            # Preserve transparency for PNG
            if im.mode in ("RGBA", "P"):
                im = im.convert("RGBA")
            else:
                im = im.convert("RGB")
            dst.parent.mkdir(parents=True, exist_ok=True)
            im.save(dst, "WEBP", quality=WEBP_QUALITY, method=6)
        return str(dst.relative_to(ROOT))
    except Exception as e:
        print(
            json.dumps({"error": f"Failed to convert {src}: {e}"}),
            file=sys.stderr,
        )
        return None


def scan_and_convert():
    converted = []

    # Public mascot PNGs
    mascot_dir = ROOT / "assets" / "images" / "mascots"
    if mascot_dir.exists():
        for src in sorted(mascot_dir.glob("*.png")):
            if "thumb" in src.name:
                continue
            result = convert_image(src)
            if result:
                converted.append(result)

    # Site assets (logolockup, logomark, mascot)
    sites_dir = ROOT / "sites"
    if sites_dir.exists():
        for ext in ("*.png", "*.jpg", "*.jpeg"):
            for src in sorted(sites_dir.glob(f"*/assets/{ext}")):
                # Merch composites are print-only masters (never referenced in
                # HTML) — multi-GB canvases that must not be re-encoded here.
                if "_comp-" in src.name:
                    continue
                result = convert_image(src)
                if result:
                    converted.append(result)

    # Root page referenced images
    for ext in ("*.png", "*.jpg", "*.jpeg"):
        for src in sorted((ROOT / "assets" / "images").glob(ext)):
            result = convert_image(src)
            if result:
                converted.append(result)

    # Brand kit assets (page visuals, ornaments, badges, seals) — plan §7.1
    brand_dir = ROOT / "assets" / "brand"
    if brand_dir.exists():
        for ext in ("*.png", "*.jpg", "*.jpeg"):
            for src in sorted(brand_dir.rglob(ext)):
                result = convert_image(src)
                if result:
                    converted.append(result)

    # OG default SVG -> PNG/WebP
    converted.extend(render_og_default())

    return converted


def main():
    converted = scan_and_convert()
    # Normalize Windows paths to forward slashes for JSON consumers.
    converted = [c.replace("\\", "/") for c in converted]
    print(json.dumps({"converted": converted, "count": len(converted)}))


if __name__ == "__main__":
    main()
